WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift

```diff
@@ -166,6 +167,30 @@ private extension ProductFormViewController {
 
         navigationController?.present(inProgressViewController, animated: true, completion: nil)
 
+        let group = DispatchGroup()
+
+        // Waits for all product images to be uploaded before updating the product remotely.
+        group.enter()
+        let observationToken = productImageActionHandler.addUpdateObserver(self) { [weak self] (productImageStatuses, error) in
+            guard productImageStatuses.hasImagesPendingUpload == false else {
+                return
+            }
+
+            guard let self = self else {
+                return
+            }
+
+            self.product = self.productUpdater.imagesUpdated(images: productImageStatuses.images)
+            group.leave()
+        }
+
+        group.notify(queue: .main) { [weak self] in
+            observationToken.cancel()
+            self?.updateProductRemotely()
+        }
+    }
```

Comment:
⚠️  I feel like this **and** `updateProductRemotely` should be in the `ViewModel`. For now, how do you feel about placing this inside a separate function like:

```swift
func waitUntilAllImagesAreUploaded(_ completion: () -> ()) {

}

func dispatchUpdateProductAction() {
}

func uploadProductRemotely() {
    waitUntilAllImagesAreUploaded {
        dispatchUpdateProductAction()
    }
}
```

Or something like that. 😄 

----

This is nice clean code by the way! I like that `group.notify()` shows what will happen _next_.

Reply:
yea the logic in `ProductFormViewController` is certainly getting out of hand and these user interactions handling logic could be in the view model! updated the function names in https://github.com/woocommerce/woocommerce-ios/pull/2126/commits/964615ce07f56dd6cb2b3d79d80ebcdff0acbc75 for now.  I'm planning on taking on https://github.com/woocommerce/woocommerce-ios/issues/1985 toward the end of Products M2 or even sooner after this PR

Reply:
Fine by me! We don't need to refactor all the time and not hit product goals. 🙂 Thanks for making the change!